src/sources/mod.rs

```diff
@@ -157,25 +164,28 @@ pub trait EventSource {
     /// [`Poll::unregister`](crate::Poll#method.unregister) method.
     fn unregister(&mut self, poll: &mut Poll) -> crate::Result<()>;
 
-    /// Notification that a polling session is going to start
+    /// Whether this source needs to be sent the [`EventSource::before_sleep`]
+    /// and [`EventSource::before_handle_events`] notifications. These are opt-in because
+    /// they require more expensive checks, and almost all sources will not need these notifications
+    const NEEDS_EXTRA_LIFECYCLE_EVENTS: bool = false;
+    /// Notification that a single `poll` is about to begin.
+    /// If this returns Ok(Some), polling will shortcircuit, and
+    /// your event handler will be called with the returned Token and Readiness
     ///
-    /// You can generate events from this method as you would from `process_events`.
-    fn pre_run<F>(&mut self, _callback: F) -> crate::Result<()>
-    where
-        F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
-    {
-        Ok(())
-    }
-
-    /// Notification that the current polling session ended
+    /// Use this to perform operations which must be done before polling,
+    /// but which may conflict with other event handlers. For example,
+    /// if polling requires a lock to be taken.
+    fn before_sleep(&mut self) -> crate::Result<Option<(Readiness, Token)>> {
+        Ok(None)
+    }
+    /// Notification that polling is completed, and the resulting events are
+    /// going to be executed.
     ///
-    /// You can generate events from this method as you would from `process_events`.
-    fn post_run<F>(&mut self, _callback: F) -> crate::Result<()>
-    where
-        F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
-    {
-        Ok(())
-    }
+    /// Use this to perform a cleanup before event handlers with arbitrary
+    /// code may run. This could be used to drop a lock obtained in
+    /// [`EventSource::before_sleep`]
+    #[allow(unused_variables)]
+    fn before_handle_events(&mut self, was_awoken: bool) {}
```

Comment:
I'm wondering if it might be relevant to pass the `(Token, Readiness)` couple as argument to this, instead of just a `was_awoken` bool. Again as a way to properly handle composite sources that may have multiple underlying FDs and tokens.

Reply:
To be more precise, I see two ways to achieve this:

- make the function take as argument `Option<(Token, Readiness)>`, and say that it is called once with `None` if no event is received, or once ore more times with `Some()` if events were triggered
- make the function take a `&dyn Iterator<Item=(Token, Readiness)>>` as argument, with that iterator being empty if no event was triggered